---
title: "CAAPA GWAS+ADPC imputation pipeline"
output: html_document
---

## Imputation summary statistics

## Folder structure

### data/raw

- <i>10411manifest.xlsx</i>: Rich's manifest file, obtained from Nick
- per site, the fam and bim files from Nick's Box account, which he originally received from the investigators; used to determine which GWAS files to use as input
- <i>caapa</i>: The CAAPA Oct 2015 reference panel, downloaded from Nick's Box account

### data/input

- <i>manifest_master.txt</i>: The first sheet of <i>10411manifest.xlsx</i> (copied and pasted)
- <i>manifest_good_sample_list.txt</i>: List of samples extracted from the first column of the GOOD worksheet of <i>10411manifest.xlsx</i>
- <i>adpc.*</i>: PLINK files for all samples typed on ADPC (originally called <i>ADPC_final</i> and just renamed)
- <i>ADPC\_SNPs_flagged_markers.txt</i>: List of SNPs that Nick and Rich identified as failing clustering
- <i>asw\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the 45 African American thousand genome individuals (ASW population). PLINK was used to calculate the minor/major allele and MAF (see section <i>5.7 Allele frequency</i> of the manual, where they clearly state that A1 = minor allele and A2 = major allele), from TGP files originally extracted for my "tgp" project. (Initially I thought to use this for determining whether I should flip ADPC ambiguous SNPs, but we decided it is better to use the CAAPA reference panel for this; however, these files might be useful for later QC troubleshooting of allele frequencies, so are retained for now.)
- <i>caapa\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the CAAPA reference panel. These files were created using the <i>create\_caapa\_freq\_files.sh</i> script, that implemented SNP QC and related indidvidual removal as documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i>. As for <i>asw\_freq\_chr*.txt</i>, PLINK was used to calculate allele frequencies.

### data/input/\<site\>/

Here I had to match the original PLINK files that Nick received from the investigators, to the equivalent GWAS file that he performed QC on. The file names listed below are the original PLINK file prefixes.

- <i>chicago</i>: CAG__2010-09-17__Illumina_1Mv1__hg19 (fam file line count = Nick's previous QC fam file line count;  bim file line count is in the same order as Nick's previous QC bim file)


## Flow diagram nomenclature

- "good" samples are those samples that are listed in <i>manifest_good_sample_list.txt</i>
- n = sample size, m = number of SNPs

## Technical notes

- When the initial ADPC and GWAS files are created, sex and phenotype values in the fam files are set to missing (0)
- The SNP positions of the initial ADPC file was fixed to match dbSNP 142 chromosome end positions (instead of the start position), but not for SNPs with rs IDs - these SNP positions were already correct. This does result in some duplicate SNPs by position (2 SNP pairs), which were removed (4 SNPs).
- The following PLINK parameters were used for the SNP QC (both GWAS and ADPC): <code>--maf 0.0001 --geno 0.05 --hwe 0.0001</code>
- For SNPs that are discordant between ADPC and GWAS (same SNP in the ADPC and GWAS data sets, for the same sample, with different genotype calls): if a SNP is discordant in 5% of samples or more, the SNP is deleted from the dataset, else the SNP value is set to missing for the sample(s) it is discordant for
- For ADPC files, there are about 70 000 strand ambiguous SNPs, therefore these SNPs cannot be simply deleted. SNPs with MAF >= 0.3 in the ADPC/GWAS or CAAPA reference panel are deleted. SNPs that cannot be mapped to the CAAPA reference panel are also deleted. SNPs with MAF < 0.3 that have a different minor allele as the CAAPA reference SNP is flipped.Note that as far as I can tell, the imputation server does not attempt to flip strand ambiguous SNPs -  from http://csg.sph.umich.edu/abecasis/MACH/tour/imputation.html: <i>"If your sample does include A/T and G/C SNPs, you'll have to ensure they are aligned to the same strand manually and inspect allele frequency discrepancies identified by MACH to help pinpoint problems. Although it is typical that a small number of SNPs will drift in frequency between populations, we recommend that you read through the warnings generated by MACH. If you see large frequency discrepancies or anything else suspicious ... investigate"</i>

## Issues

- For Chicago, there is one additional GOOD common sample, WG0237021-DNAA02, which was not inclduded in the original run. I don't know why...(In the Master sheet of the manifest, all flags are 0 for this sample, it is in the GOOD sheet, it is NOT in the illumina fails sheet, is is NOT in the QC errors sheet.)
- When calculating CAAPA allele frequencies, the related individuals documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i> are first removed. However, the following 3 IDs that are listed for removal, and is supposed to be in the VCF files, are not: 330119, 750200001, 750200047. 