---
title: "CAAPA GWAS+ADPC imputation pipeline"
output: word_document
---

## Summary statistics

### Pre-imputation QC

```{r, echo=FALSE}
source("functions.R")
knitr::kable(getPreQCStatistics(), align=c("l", rep("r",10)))
```

### Imputation QC

These statistics were scraped from the imputation server's QC reports and summarized into tables, so that statistics can be compared across sites.

#### ADPC

```{r, echo=FALSE}
#source("functions.R")
#knitr::kable(getImputationQCStatistics1("adpc"), align=c("l", rep("r",7)))
#knitr::kable(getImputationQCStatistics2("adpc"), align=c("l", rep("r",6)))
```

#### GWAS

```{r, echo=FALSE}
#source("functions.R")
#knitr::kable(getImputationQCStatistics1("gwas"), align=c("l", rep("r",7)))
#knitr::kable(getImputationQCStatistics2("gwas"), align=c("l", rep("r",6)))
```

### Merged

```{r, echo=FALSE}
source("functions.R")
#knitr::kable(getImputationQCStatistics1("imputation"), align=c("l", rep("r",7)))
#knitr::kable(getImputationQCStatistics2("imputation"), align=c("l", rep("r",6)))
```

### Post-imputation QC

SNPs were deleted per site based on the following metric:
SNPs with MAF <= 0.005 were deleted if Rsq <= 0.5;
SNPs with MAF > 0.005 were deleted if Rsq <= 0.3;

This was based on http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3410659/, the paragraph just above the heading "TO INCLUDE OR NOT TO INCLUDE: RARE SNPs DURING HAPLOTYPE RECONSTRUCTION".

#### Median Rsq

The Rsq distribution is not normal, which is why the median was chosen as centrality statistic.
```{r, echo=FALSE}
source("functions.R")
drawMedianRsq()
```


#### Proportion SNPs deleted accoss the genome

```{r, echo=FALSE}
source("functions.R")
#knitr::kable(getSNPsDeleted(), align=c("l", rep("r",9)))
```

#### Proportion rare SNPs (MAF <= 0.005) deleted per chromosome

```{r, echo=FALSE}
source("functions.R")
#drawSNPsDeletedHeatMap("lt")
```

#### Proportion other SNPs (MAF > 0.005) deleted per chromosome

```{r, echo=FALSE}
source("functions.R")
#drawSNPsDeletedHeatMap("gt")
```

#### Proportion SNPs with large frequency difference (> 0.05) compared to reference across the genome

```{r, echo=FALSE}
source("functions.R")
#getSNPsDiffStats() #This takes long to run - uncomment only if necessary to regenerate data/output/*diff*.txt files
#knitr::kable(getGenomeFreqDiff(read.table("../data/output/ref_diff_05.txt", head=T)), align=c("l", rep("r",3))) 
```

#### Proportion SNPs with large frequency difference (> 0.1) compared to reference across the genome

```{r, echo=FALSE}
source("functions.R")
#knitr::kable(getGenomeFreqDiff(read.table("../data/output/ref_diff_10.txt", head=T)), align=c("l", rep("r",3))) 
```

#### Proportion SNPs with large frequency difference (> 0.1) compared to reference per chromosome

```{r, echo=FALSE}
source("functions.R")
#drawFreqDiffHeatMap(read.table("../data/output/ref_diff_10.txt", head=T))
```

#### Proportion genotyped SNPs with large frequency difference (> 0.1) compared to imputed version across the genome

```{r, echo=FALSE}
source("functions.R")
#knitr::kable(getGenomeFreqDiff(read.table("../data/output/geno_diff_10.txt", head=T)), align=c("l", rep("r",3))) 
```

#### Proportion genotyped SNPs with large frequency difference (> 0.1) compared to imputed version per chromosome

```{r, echo=FALSE}
source("functions.R")
#drawFreqDiffHeatMap(read.table("../data/output/geno_diff_10.txt", head=T))
```

## Nomenclature

- n = sample size, m = number of SNPs
- disconcordant samples: samples with the same ID that are not IBD duplicates in the ADPC and GWAS data set

## Flow diagram nomenclature

- "good" samples are those samples that are listed in <i>manifest_good_sample_list.txt</i>

## Technical notes

- All the sites were run through on git revision b5e4c15, with the exception of jhu_650y, which was run on 72bb543 (new file pre-processing introduced, since we realized that a lot of SNPs were missing from the original file that Hiro prepared)
- When the initial ADPC and GWAS files are created, sex and phenotype values in the fam files are set to missing (0)
- The SNP positions of the initial ADPC file was fixed to match dbSNP 142 chromosome end positions (instead of the start position), but not for SNPs with rs IDs - these SNP positions were already correct. This does result in some duplicate SNPs by position (2 SNP pairs), which were removed (4 SNPs).
- For some sites, there are a small number of duplicate SNPs by position in the GWAS dataset. All such SNPs are removed. 
- The following PLINK parameters were used for the SNP QC (both GWAS and ADPC): <code>--maf 0.0001 --geno 0.05 --hwe 0.0001</code>. It may be a bit hard to justify this HWE threshold since 10^(-6) is what is normally used. 10^(-4) was the setting that Nick originally in his pipeline documentation recommended, and I carried it forward. To me it makes sense to use a slightly stringent threshold as there seemed to have been a lot of quality issues with the ADPC chip. The following link also uses this threshold: http://hrsonline.isr.umich.edu/sitedocs/genetics/HRS\_QC\_REPORT\_MAR2012.pdf, and a reference to support it: http://www.ncbi.nlm.nih.gov/pubmed/18217698
- For SNPs that are discordant between ADPC and GWAS (same SNP in the ADPC and GWAS data sets, for the same sample, with different genotype calls): if a SNP is discordant in 5% of samples or more, the SNP is deleted from the dataset, else the SNP value is set to missing for the sample(s) it is discordant for
- For the imputation configuration, the <i>CAAPA - African American</i> reference panel was used, and the configured population was <i>AA (CAAPA)</i>
- When calculating CAAPA allele frequencies, the related individuals documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i> are first removed. The remaining 883 samples are exactly the same 883 samples that we sent as the CAAPA reference panel to Michigan (checked against <i>data/raw/caapa/CAAPA_chr22_manifest_postFiltering.txt</i>)
- For Chicago, there is one additional GOOD common sample, WG0237021-DNAA02, which was not inclduded in the original run. I don't know why...(In the Master sheet of the manifest, all flags are 0 for this sample, it is in the GOOD sheet, it is NOT in the illumina fails sheet, and NOT in the QC errors sheet.). Nick confirmed that there is no reason to remove this sample, and it will be included in the analysis.
- For the JHU 650Y and BDOS sites, there are no AT/CG SNPs in the GWAS data sets, since the array did not capture any AT/CG SNPs (according to Nick).
- A manual check was done to ensure that there are no IBD duplicates in the GWAS QC data set; no IBD duplicates were found in any of the sites checked 
- A check was also done to ensure that there are no discordant ADPC + GWAS samples (sample with the same ID in the ADPC and GWAS data sets, that are NOT IBD duplicates): if a site has e.g. 1000 samples, 10 samples could theoretically be discordant, and overall concordance between the data sets could then still be 99\%. With the exception of Jackson JHS, JHU ABR, UCSF SanFrancisco and Winston Salem, discordant samples were found for all of the sites. There are no crossmatched samples (sample with different IDs in the ADPC and GWAS data sets, that are IBD duplicates), except for Detroit, where there is 1 pair of crossmatched samples, which Nick and I tried to resolve, but could not. Discordant samples were deleted from the data sets.
- For ADPC files, there are about 70 000 strand ambiguous SNPs, therefore these SNPs cannot be simply deleted. AT/CG SNPs are handled as follows: SNPs not in the reference panel are deleted; SNPs with MAF > 0.4 in either the data set or reference panel are deleted; SNPs that differ in MAF more than 1\% compared to the reference panel are deleted; for the remaining SNPs, if the minor allele differs, the strand of the SNP is flipped. These thresholds are stringent and a large nr of SNPs are deleted; however the impact that incorrect AT/CG SNPs may have on imputation could be severe, so it is better to use stringent thresholds. Note that as far as I can tell, the imputation server does not attempt to flip strand ambiguous SNPs -  from http://csg.sph.umich.edu/abecasis/MACH/tour/imputation.html: <i>"If your sample does include A/T and G/C SNPs, you'll have to ensure they are aligned to the same strand manually and inspect allele frequency discrepancies identified by MACH to help pinpoint problems. Although it is typical that a small number of SNPs will drift in frequency between populations, we recommend that you read through the warnings generated by MACH. If you see large frequency discrepancies or anything else suspicious ... investigate"</i>
- Proportionally, a lot more AT/CG SNPs are deleted for Barbados, ABR, and Puerto Rico, since these populations may be distinct from African Americans. We should perhaps have considered changing the MAF cut-off threshold for these sites, but in order to be consistent across sites, we didn't.
- For UCSF PR, there are 4 additional GOOD common samples(WG0233330-DNAH04\_HR5371, WG0233320-DNAA02\_NY50126, WG0233342-DNAC01\_SJ1252, WG0233354-DNAF02\_SJ5348), which was not inclduded in the original run. I don't know why...(In the Master sheet of the manifest, all flags are 0 for these samples, they are in the GOOD sheet, and NOT in the illumina fails sheet, and NOT in the QC errors sheet.). 
- For the JHU sites, the initial GWAS file has samples with more than 5\% missing genotypes. These samples were removed during the GWAS QC step. For Detroit, there were 2 samples with borderline missingness (WG0233409-DNAH10\_11593-9 and WG0233409-DNAE11\_11608-3, missingness 0.05121 and 0.05182). Since this was so borderline (which was not the case for the JHU sites), I decided to retain these samples.
- The Jackson site was genotyped in 2 different batches, the one having better quality than the other (the JHS samples performed better than the ARIC samples). In order to not remove SNPs unnecessarily from the “JHS” batch, the batches were QC’ed and imputed separately 

### File pre-processing steps

### Detroit

This site required some pre-processing since its bim file encoded SNPs as 1/2 instead of A/T/C/G. 
See script <i>preprocess_detroit.sh</i>.

#### Jackson

The GWAS bim file was lifted over from hg18 to hg19, using CrossMap (http://crossmap.sourceforge.net), and the hg18tohg19 chain file published with this software. In addition, the ARIC and JHS sites had to be split out from the original file, according to <i>data/input/jackson/site_map.txt</i>

### JHU sites

It is my understanding that the non-JHU sites did their own GWAS QC at a sample level; looking at the maximum percentage of missing genotypes per sample per site, these sites appeared to use a threshold of 5\%. The JHU sites appeared to use a much higher threshold, which Nick explained was due to historic reasons. In order to be consistent, samples with more than 5\% missing genotypes were removed in the GWAS QC step (plink parameters <code>--mind 0.05</code>) for each of the 3 JHU sites, resulting in the loss of 2, 1, and 9 samples in the JHU 650Y, JHU ABR and JHU BDOS sites respectively.

### JHU ABR

Some sample swaps occurred for this site in the GWAS data set. The sample identities were corrected according to <i>data/input/jhu_abr/sample\_delete.txt</i> (old ID, new ID; 18 samples), and 4 samples were deleted <i>data/input/jhu\_abr/sample\_delete.txt</i> (502001034, 502001069, 502001149, 502001136). The last sample was deleted due to it being an IBD duplicate of another sample in the data set, 502001135 (missing rate for 502001136 = 2118/1992457 in the GWAS QC data set, missing rate for 502001135 = 2019/1992457 in the GWAS QC data set, i.e. 502001136 had the highest missing rate). Both of these samples were discordant with their corresponding ADPC sample ID, and instead was IBD duplicates of ADPC sample WG0233367-DNAA04.

### JHU Barbados

Samples with suspect identities identified for the Barbados sample identity project were deleted from the GWAS file (<i>data/input/jhu\_barbados/sample\_delete.txt</i>; 9 samples).

### JHU 650y

The coordinates of this data set was lifted over using the JHU Barbados BIM file. 


### Washington 

The following samples are IBD duplicates, but for some reason do not show up in the <i>mz dups</i> worksheet of the manifest: WG0237441-DNAE02 and WG0237461-DNAH12; WG0237458-DNAD06 and WG0237462-DNAF09. In addition, the following 2 samples were swapped around in either the ADPC or GWAS data sets:

- ADPC_WG0237446-DNAD04 GWAS_WG0237450-DNAC03 PI_HAT=0.99
- ADPC_WG0237450-DNAC03 GWAS_WG0237446-DNAD04 PI_HAT=1.00

Nick followed up with Ayo, who said that they removed all 6 these samples due to inconsistencies. These 6 samples will therefore be deleted (<i>data/input/washingon/sample\_delete.txt</i> )

## Issues


## Folder structure

### data/raw

- <i>10411manifest.xlsx</i>: Rich's manifest file, obtained from Nick
- per site, the fam and bim files from Nick's Box account, which he originally received from the investigators; used to determine which GWAS files to use as input
- <i>caapa</i>: The CAAPA Oct 2015 reference panel, downloaded from Nick's Box account.

### data/input

- <i>manifest_master.txt</i>: The first sheet of <i>10411manifest.xlsx</i> (copied and pasted; "JHS / Uvermont" was changed to JHS\_Uvermont, to avoid institution matching script issues
- <i>manifest_good_sample_list.txt</i>: List of samples extracted from the first column of the GOOD worksheet of <i>10411manifest.xlsx</i>, that has <i>Remove</i> set to 0. The following 4 additional samples were added manually to the list: WG0233367-DNAA04, WG0233368-DNAG03, WG0233368-DNAC01, WG0233367-DNAD12. These samples are JHU ABR samples that were mislabelled with different sample IDs in the GWAS data set, but their identities were since corrected. (These 4 samples have <i>Remove</i> set to 1 since they had no GWAS data; I checked that there are no other ABR samples that were also marked for removal also for not having GWAS data; these were the only 4 samples.)
- <i>adpc.*</i>: PLINK files for all samples typed on ADPC (originally called <i>ADPC_final</i> and just renamed)
- <i>ADPC\_SNPs_flagged_markers.txt</i>: List of SNPs that Nick and Rich identified as failing clustering
- <i>asw\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the 45 African American thousand genome individuals (ASW population). PLINK was used to calculate the minor/major allele and MAF (see section <i>5.7 Allele frequency</i> of the manual, where they clearly state that A1 = minor allele and A2 = major allele), from TGP files originally extracted for my "tgp" project. (Initially I thought to use this for determining whether I should flip ADPC ambiguous SNPs, but we decided it is better to use the CAAPA reference panel for this; however, these files might be useful for later QC troubleshooting of allele frequencies, so are retained for now.)
- <i>caapa\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the CAAPA reference panel. These files were created using the <i>create\_caapa\_freq\_files.sh</i> script, that implemented SNP QC and related individual removal as documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i>. As was done for <i>asw\_freq\_chr*.txt</i>, PLINK was used to calculate allele frequencies. (Note that these files may contain a FILTER field of "."" instead pf PASS for some SNPs even though their quality is OK; this due to a bug when the VCF files were originally merged from the single sample VCF files. The SNP filtering applied was therefore done EXACTLY as documented in the VCF release document - initially the flag <i>remove-filtered-all</i> was also specified but that resulted in a fair proportion of SNPs being deleted erroneously, which means that a large nr of AT/CG SNPs were removed unncessarily.)

### data/input/\<site\>/

Here I had to match the original PLINK files that Nick received from the investigators, to the equivalent GWAS file that he performed QC on. The file names listed below are the original PLINK file prefixes.

- <i>chicago</i>: CAG__2010-09-17__Illumina_1Mv1__hg19 (fam file line count = Nick's previous QC fam file line count;  bim file line count is in the same order as Nick's previous QC bim file)
- <i>detroit</i>: SAPPHIRE_Detroit_GWAS/sapphire\_adpcNDNC 
- <i>jackson</i>: JHS/JHS\_CAAPA\_AFFY6; site\_map.txt was extracted from JHS\_992GenotypedSamples\_ForCAAPA_Phenotypes\_15May2015.xls (received in an email from Nick)
- <i>jhu\_650y</i>: Extracted from BC gene (the BOX folder GRAAD/GRAAD\_650Y had ~ 185 000 SNPs to little; originally this file was used in the pipeline, but we then noticed the discrepancy in SNP numbers)
- <i>jhu\_abr</i>: GRAAD/ABR\_GWAS; files for GWAS sample identity fixes are also in this folder
- <i>jhu\_bdos</i>: Extracted from BC gene; a file containing a list of samples with suspect IDs that should be deleted also located in this folder
- <i>ucsf\_pr</i>: Puerto Rico/GALA2\_hg19\_NA34\_121214\_rsid\_PuertoRicans\_ADPC
- <i>ucsf\_sf</i>: UCSF/SAGE\_hg19\_NA3\4_082814\_rsid\_AfricanAmericans\_ADPC
- <i>washington</i>: Rotimi/hgc\_all\_501\_autosome\_b37\_d8 
- <i>winston\_salem</i>: WFU/SARP\_ADPC (note that SAD is entirely contained in SARP)

### data/output/\<site\>/

- merged: the GWAS+ADPC merged data sets used for imputation
- imputed: imputation output files
- flow: flow diagram of the pipeline run for the site
- statistics: the imputation server summary reports and statistic files for the GWAS QC, ADPC QC, and imputation