---
title: "CAAPA GWAS+ADPC imputation pipeline"
output: word_document
---

## Imputation summary statistics


## Flow diagram nomenclature

- "good" samples are those samples that are listed in <i>manifest_good_sample_list.txt</i>
- n = sample size, m = number of SNPs

## Technical notes

- When the initial ADPC and GWAS files are created, sex and phenotype values in the fam files are set to missing (0)
- The SNP positions of the initial ADPC file was fixed to match dbSNP 142 chromosome end positions (instead of the start position), but not for SNPs with rs IDs - these SNP positions were already correct. This does result in some duplicate SNPs by position (2 SNP pairs), which were removed (4 SNPs).
- For some sites, there are a small number of duplicate SNPs by position in the GWAS dataset. All such SNPs are removed. 
- The following PLINK parameters were used for the SNP QC (both GWAS and ADPC): <code>--maf 0.0001 --geno 0.05 --hwe 0.0001</code>
- For SNPs that are discordant between ADPC and GWAS (same SNP in the ADPC and GWAS data sets, for the same sample, with different genotype calls): if a SNP is discordant in 5% of samples or more, the SNP is deleted from the dataset, else the SNP value is set to missing for the sample(s) it is discordant for
- For ADPC files, there are about 70 000 strand ambiguous SNPs, therefore these SNPs cannot be simply deleted. SNPs with MAF differences > 0.05 between the ADPC/GWAS and CAAPA reference panel are deleted. SNPs that cannot be mapped to the CAAPA reference panel are also deleted. The remaining SNPs are flipped if the minor allele is different between the ADPC/GWAS data set and CAAPA reference panel. Note that as far as I can tell, the imputation server does not attempt to flip strand ambiguous SNPs -  from http://csg.sph.umich.edu/abecasis/MACH/tour/imputation.html: <i>"If your sample does include A/T and G/C SNPs, you'll have to ensure they are aligned to the same strand manually and inspect allele frequency discrepancies identified by MACH to help pinpoint problems. Although it is typical that a small number of SNPs will drift in frequency between populations, we recommend that you read through the warnings generated by MACH. If you see large frequency discrepancies or anything else suspicious ... investigate"</i>
- For the imputation configuration, the <i>CAAPA - African American</i> reference panel was used, and the configured population was <i>AA (CAAPA)</i>
- When calculating CAAPA allele frequencies, the related individuals documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i> are first removed. The remaining 883 samples are exactly the same 883 samples that we sent as the CAAPA reference panel to Michigan (checked against <i>data/raw/caapa/CAAPA_chr22_manifest_postFiltering.txt</i>)
- For Chicago, there is one additional GOOD common sample, WG0237021-DNAA02, which was not inclduded in the original run. I don't know why...(In the Master sheet of the manifest, all flags are 0 for this sample, it is in the GOOD sheet, it is NOT in the illumina fails sheet, and NOT in the QC errors sheet.). Nick confirmed that there is no reason to remove this sample, and it will be included in the analysis.
- For JHU 650Y, there are no AT/CG SNPs in the GWAS data set with MAF that are discrepant with the CAAPA reference panel. I double checked that there are indeed some AT/CG SNPs in this dataset (to make sure that this was not removed pre-QC as was originally the case for Detroit: there are 3283.

### File pre-processing steps

#### Jackson

The GWAS bim file was lifted over from hg18 to hg19, using CrossMap (http://crossmap.sourceforge.net), and the hg18tohg19 chain file published with this software. In addition, the ARIC and JHS sites had to be split out from the original file, according to <i>data/input/jackson/site_map.txt</i>

### JHU ABR

Some sample swaps occurred for this site in the GWAS data set. The sample identities were corrected according to <i>data/input/jhu_abr/sample\_delete.txt</i> (old ID, new ID; 14 samples), and 3 samples were deleted <i>data/input/jhu\_abr/sample\_delete.txt</i> (502001034, 502001069, 502001149).

### JHU Barbados

Samples with suspect identities identified for the Barbados sample identity project were deleted from the GWAS file (<i>data/input/jhu\_barbados/sample\_delete.txt</i>; 9 samples).

## Issues
- On average, there are 5000 ADPC AT/CG SNPs that have a MAF difference of more than 2.5% compared to the CAAPA reference panel. For Barbados and ABR, this is about double (roughly 10 000 SNPs), and for Puerto Rico, this is almost 5 times more (roughly 24 000 SNPs). Since these populations may be distinct from African Americans, should we change this threshold to 5% for these sites?
- For UCSF PR, there are 4 additional GOOD common samples(WG0233330-DNAH04\_HR5371, WG0233320-DNAA02\_NY50126, WG0233342-DNAC01\_SJ1252, WG0233354-DNAF02\_SJ5348), which was not inclduded in the original run. I don't know why...(In the Master sheet of the manifest, all flags are 0 for these samples, they are in the GOOD sheet, and NOT in the illumina fails sheet, and NOT in the QC errors sheet.). 
- In the same way as for UCSF PR, for Washington, there are 21 additional GOOD common samples: 
```{r, echo=FALSE}
prev.fam <- read.table("/Volumes/Promise Pegasus/caapa_chip_data_merging/iteration_1/GWAS_files/Washington/GWAS/NIH_keep.fam")
new.fam <- read.table("/Volumes/Promise Pegasus/caapa_imputation_pipeline/data/working/washington/gwas_common.fam")
#dim(prev.fam)
#dim(new.fam)
#sum(prev.fam$V2 %in% new.fam$V2)
new.fam[!(new.fam$V2 %in% prev.fam$V2),2]
```


## Folder structure

### data/raw

- <i>10411manifest.xlsx</i>: Rich's manifest file, obtained from Nick
- per site, the fam and bim files from Nick's Box account, which he originally received from the investigators; used to determine which GWAS files to use as input
- <i>caapa</i>: The CAAPA Oct 2015 reference panel, downloaded from Nick's Box account.

### data/input

- <i>manifest_master.txt</i>: The first sheet of <i>10411manifest.xlsx</i> (copied and pasted; "JHS / Uvermont" was changed to JHS\_Uvermont, to avoid institution matching script issues
- <i>manifest_good_sample_list.txt</i>: List of samples extracted from the first column of the GOOD worksheet of <i>10411manifest.xlsx</i>
- <i>adpc.*</i>: PLINK files for all samples typed on ADPC (originally called <i>ADPC_final</i> and just renamed)
- <i>ADPC\_SNPs_flagged_markers.txt</i>: List of SNPs that Nick and Rich identified as failing clustering
- <i>asw\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the 45 African American thousand genome individuals (ASW population). PLINK was used to calculate the minor/major allele and MAF (see section <i>5.7 Allele frequency</i> of the manual, where they clearly state that A1 = minor allele and A2 = major allele), from TGP files originally extracted for my "tgp" project. (Initially I thought to use this for determining whether I should flip ADPC ambiguous SNPs, but we decided it is better to use the CAAPA reference panel for this; however, these files might be useful for later QC troubleshooting of allele frequencies, so are retained for now.)
- <i>caapa\_freq\_chr\*.txt</i>: CHR, SNP, POS, A1 (minor allele),	A2 (major allele), MAF, for the CAAPA reference panel. These files were created using the <i>create\_caapa\_freq\_files.sh</i> script, that implemented SNP QC and related individual removal as documented in <i>VCF\_DataRelease\_README\_Oct2015.pdf</i>. As was done for <i>asw\_freq\_chr*.txt</i>, PLINK was used to calculate allele frequencies.

### data/input/\<site\>/

Here I had to match the original PLINK files that Nick received from the investigators, to the equivalent GWAS file that he performed QC on. The file names listed below are the original PLINK file prefixes.

- <i>chicago</i>: CAG__2010-09-17__Illumina_1Mv1__hg19 (fam file line count = Nick's previous QC fam file line count;  bim file line count is in the same order as Nick's previous QC bim file)
- <i>jackson</i>: JHS/JHS\_CAAPA\_AFFY6; site\_map.txt was extracted from JHS\_992GenotypedSamples\_ForCAAPA_Phenotypes\_15May2015.xls (received in an email from Nicl)
- <i>jhu\_650y</i>: GRAAD/GRAAD\_650Y
- <i>jhu\_abr</i>: GRAAD/ABR\_GWAS; files for GWAS sample identity fixes are also in this folder
- <i>jhu\_bdos</i>: Extracted from BC gene; a file containing a list of samples with suspect IDs that should be deleted also located in this folder
- <i>ucsf\_pr</i>: Puerto Rico/GALA2\_hg19\_NA34\_121214\_rsid\_PuertoRicans\_ADPC
- <i>ucsf\_sf</i>: UCSF/SAGE\_hg19\_NA3\4_082814\_rsid\_AfricanAmericans\_ADPC
- <i>washington</i>: Rotimi/hgc\_all\_501\_autosome\_b37\_d8 
- <i>winston\_salem</i>: WFU/SARP\_ADPC (note that SAD is entirely contained in SARP)

### data/output/\<site\>/

- merged: the GWAS+ADPC merged data sets used for imputation
- imputed: imputation output files
- flow: flow diagram of the pipeline run for the site
- statistics: the imputation server summary reports and statistic files for the GWAS QC, ADPC QC, and imputation